pkg/frontend/querymiddleware/model_extra.go

```diff
@@ -331,32 +336,67 @@ func (vs *vectorSampleStream) UnmarshalJSON(b []byte) error {
 }
 
 func (vs vectorSampleStream) MarshalJSON() ([]byte, error) {
-	if len(vs.Samples) != 1 {
-		return nil, fmt.Errorf("vector sample stream should have exactly one sample, got %d", len(vs.Samples))
+	if (len(vs.Samples) == 1) == (len(vs.Histograms) == 1) { // not XOR
+		return nil, fmt.Errorf("vector sample stream should have exactly one sample or one histogram, got %d samples and %d histograms", len(vs.Samples), len(vs.Histograms))
 	}
-	return json.Marshal(model.Sample{
-		Metric:    mimirpb.FromLabelAdaptersToMetric(vs.Labels),
-		Timestamp: model.Time(vs.Samples[0].TimestampMs),
-		Value:     model.SampleValue(vs.Samples[0].Value),
-	})
+	var sample model.Sample
+	if len(vs.Samples) == 1 {
+		sample = model.Sample{
+			Metric:    mimirpb.FromLabelAdaptersToMetric(vs.Labels),
+			Timestamp: model.Time(vs.Samples[0].TimestampMs),
+			Value:     model.SampleValue(vs.Samples[0].Value),
+		}
+	} else {
+		sample = model.Sample{
+			Metric:    mimirpb.FromLabelAdaptersToMetric(vs.Labels),
+			Timestamp: model.Time(vs.Histograms[0].TimestampMs),
+			Histogram: mimirpb.FromFloatHistogramToPromHistogram(vs.Histograms[0].Histogram.ToPrometheusModel()),
+		}
+	}
+	return json.Marshal(sample)
 }
 
 // UnmarshalJSON implements json.Unmarshaler.
 func (s *SampleStream) UnmarshalJSON(data []byte) error {
 	var stream struct {
-		Metric model.Metric     `json:"metric"`
-		Values []mimirpb.Sample `json:"values"`
+		Metric     model.Metric                  `json:"metric"`
+		Values     []mimirpb.Sample              `json:"values"`
+		Histograms []mimirpb.SampleHistogramPair `json:"histograms"`
 	}
 	if err := json.Unmarshal(data, &stream); err != nil {
 		return err
 	}
 	s.Labels = mimirpb.FromMetricsToLabelAdapters(stream.Metric)
-	s.Samples = stream.Values
+	if len(stream.Values) > 0 {
+		s.Samples = stream.Values
+	}
+	if len(stream.Histograms) > 0 {
+		return fmt.Errorf("cannot unmarshal native histograms from JSON, but stream contains %d histograms", len(stream.Histograms))
+	}
 	return nil
 }
 
 // MarshalJSON implements json.Marshaler.
 func (s *SampleStream) MarshalJSON() ([]byte, error) {
```

Comment:
It should be fine to have `values` and `histograms` in the same entry. If you are referring Prometheus docs, the docs [here](https://prometheus.io/docs/prometheus/latest/querying/api/#expression-query-result-formats) are a little outdated. Please see [this](https://github.com/prometheus/prometheus/blob/main/docs/querying/api.md#range-vectors) (the website will get updated with the next release)